// vuepress-plugin-md-enhance/lib/node/index.js
"use strict";
const path_1 = require("path");
const lineNumbers = require("@vuepress/markdown/lib/lineNumbers");
const code_demo_1 = require("./markdown-it/code-demo");
const decode_url_1 = require("./markdown-it/decode-url");
const flowchart_1 = require("./markdown-it/flowchart");
const footnote_1 = require("./markdown-it/footnote");
const katex_1 = require("./markdown-it/katex");
const mark_1 = require("./markdown-it/mark");
const mermaid_1 = require("./markdown-it/mermaid");
const presentation_1 = require("./markdown-it/presentation");
const sub_1 = require("./markdown-it/sub");
const sup_1 = require("./markdown-it/sup");
const tasklist_1 = require("./markdown-it/tasklist");
const pluginConfig_1 = require("./pluginConfig");
const mdEnhancePlugin = (option, context) => {
    const { themeConfig } = context;
    const markdownOptions = Object.keys(option).length === 0 ? themeConfig.mdEnhance || {} : option;
    const alignEnable = markdownOptions.enableAll || markdownOptions.align || false;
    const demoEnable = markdownOptions.enableAll || markdownOptions.demo || false;
    const flowchartEnable = markdownOptions.enableAll || markdownOptions.flowchart || false;
    const footnoteEnable = markdownOptions.enableAll || markdownOptions.footnote || false;
    const tasklistEnable = markdownOptions.enableAll || markdownOptions.tasklist || false;
    const mermaidEnable = markdownOptions.enableAll || Boolean(markdownOptions.mermaid) || false;
    const presentationEnable = markdownOptions.enableAll || Boolean(markdownOptions.presentation) || false;
    const texEnable = markdownOptions.enableAll || Boolean(markdownOptions.tex) || false;
    const revealPlugins = typeof markdownOptions.presentation === "object" &&
        Array.isArray(markdownOptions.presentation.plugins)
        ? markdownOptions.presentation.plugins
        : [];
    return Object.assign(Object.assign({ name: "md-enhance", alias: {
            "@FlowChart": flowchartEnable
                ? (0, path_1.resolve)(__dirname, "../client/FlowChart.vue")
                : "@mr-hope/vuepress-shared/lib/esm/noopModule",
            "@Mermaid": mermaidEnable
                ? (0, path_1.resolve)(__dirname, "../client/Mermaid.js")
                : "@mr-hope/vuepress-shared/lib/esm/noopModule",
            "@Presentation": presentationEnable
                ? (0, path_1.resolve)(__dirname, "../client/Presentation.vue")
                : "@mr-hope/vuepress-shared/lib/esm/noopModule",
        }, define: () => ({
            MARKDOWN_ENHANCE_ALIGN: alignEnable,
            MARKDOWN_ENHANCE_DELAY: markdownOptions.delay || 500,
            MARKDOWN_ENHANCE_FLOWCHART: flowchartEnable,
            MARKDOWN_ENHANCE_FOOTNOTE: footnoteEnable,
            MARKDOWN_ENHANCE_MERMAID: mermaidEnable,
            MARKDOWN_ENHANCE_PRESENTATION: presentationEnable,
            MARKDOWN_ENHANCE_TASKLIST: tasklistEnable,
            MARKDOWN_ENHANCE_TEX: texEnable,
            CODE_DEMO_OPTIONS: Object.assign(Object.assign({}, code_demo_1.codeDemoDefaultSetting), (typeof markdownOptions.demo === "boolean"
                ? {}
                : markdownOptions.demo)),
            MERMAID_OPTIONS: typeof markdownOptions.mermaid === "object"
                ? markdownOptions.mermaid
                : {},
            REVEAL_CONFIG: typeof markdownOptions.presentation === "object" &&
                typeof markdownOptions.presentation.revealConfig === "object"
                ? markdownOptions.presentation.revealConfig
                : {},
            REVEAL_PLUGIN_HIGHLIGHT: revealPlugins.includes("highlight"),
            REVEAL_PLUGIN_MATH: revealPlugins.includes("math"),
            REVEAL_PLUGIN_NOTES: revealPlugins.includes("notes"),
            REVEAL_PLUGIN_SEARCH: revealPlugins.includes("search"),
            REVEAL_PLUGIN_ZOOM: revealPlugins.includes("zoom"),
        }), enhanceAppFiles: (0, path_1.resolve)(__dirname, "../client/enhanceAppFile.js") }, (demoEnable
        ? {
            clientRootMixin: (0, path_1.resolve)(__dirname, "../client/clientRootMixin.js"),
        }
        : {})), { chainMarkdown: (md) => {
            if (markdownOptions.imageFix !== false)
                md.plugin("decode-url").use(decode_url_1.default);
            if (markdownOptions.lineNumbers !== false)
                md.plugin("line-numbers").use(lineNumbers);
            if (markdownOptions.sup || markdownOptions.enableAll)
                md.plugin("sup").use(sup_1.default);
            if (markdownOptions.sub || markdownOptions.enableAll)
                md.plugin("sub").use(sub_1.default);
            if (footnoteEnable)
                md.plugin("footnote").use(footnote_1.default);
            if (flowchartEnable)
                md.plugin("flowchart").use(flowchart_1.default);
            if (markdownOptions.mark || markdownOptions.enableAll)
                md.plugin("mark").use(mark_1.default);
            if (tasklistEnable)
                md.plugin("tasklist").use(tasklist_1.default, [
                    typeof markdownOptions.tasklist === "object"
                        ? markdownOptions.tasklist
                        : {},
                ]);
            if (mermaidEnable)
                md.plugin("mermaid").use(mermaid_1.default);
            if (texEnable)
                md.plugin("katex").use(katex_1.default, [
                    Object.assign({ macros: {
                            // support more symbols
                            "\\liiiint": "\\int\\!\\!\\!\\iiint",
                            "\\iiiint": "\\int\\!\\!\\!\\!\\iiint",
                            "\\idotsint": "\\int\\!\\cdots\\!\\int",
                        } }, (typeof markdownOptions.tex === "object"
                        ? markdownOptions.tex
                        : {})),
                ]);
            if (presentationEnable)
                md.plugin("presentation").use(presentation_1.default);
        }, plugins: (0, pluginConfig_1.getPluginConfig)(markdownOptions, context) });
};
module.exports = mdEnhancePlugin;
